### PR TITLE
Support slicing in `MultiEmbeddingTensor`

### PR DESCRIPTION
**Pull request by akihironitta**
Part of #88. Unblocks #194 and #206.

### CODE REVIEW

File: torch_frame/data/multi_embedding_tensor.py [hunk 1]
```diff
@@ -98,29 +101,159 @@ def __getitem__(
...
-            index = self._normalize_index(index, dim=0)
-            return MultiEmbeddingTensor(
-                num_rows=1,
-                num_cols=self.num_cols,
-                values=self.values[index].view(1, -1),
-                offset=self.offset,
-            )
+        if isinstance(index, tuple):
+            assert len(index) == 2
+            if isinstance(index[0], int) and isinstance(index[1], int):
```
Comment (zechengz) [inline torch_frame/data/multi_embedding_tensor.py:106]: nit
```suggestion
            if all(isinstance(x, int) for x in index):
```

File: torch_frame/data/multi_embedding_tensor.py [hunk 2]
```diff
@@ -98,29 +101,159 @@ def __getitem__(
...
+                values=torch.tensor([], device=self.device),
+                offset=torch.tensor([0], device=self.device),
+            )
+
+        elif dim == 0:
+            return MultiEmbeddingTensor(
+                num_rows=length,
+                num_cols=self.num_cols,
+                values=self.values[start:start + length],
+                offset=self.offset,
```
Comment (zechengz) [inline torch_frame/data/multi_embedding_tensor.py:218]: Should we also change the offset here?

File: torch_frame/data/multi_embedding_tensor.py [hunk 3]
```diff
@@ -98,29 +101,159 @@ def __getitem__(
...
+                out = self
+                for dim, idx in enumerate(index):
+                    out = out.select(idx, dim=dim)
+                return out
+        else:
+            return self.select(index, dim=0)
+
+    def select(
+        self,
+        index: Union[int, Tensor, List[int], slice, range],
```
Comment (yiweny) [inline torch_frame/data/multi_embedding_tensor.py:120]: Should `Tuples` be supported?

File: test/data/test_multi_embedding_tensor.py [hunk 4]
```diff
@@ -21,6 +21,29 @@ def assert_equal(
...
+    tensor_list: List[torch.Tensor],
+    index: Union[List[int], slice],
+) -> List[torch.Tensor]:
+    """Selects rows from a list of column tensors.
+
+    Args:
+        tensor_list (list[torch.Tensor]): a list of tensors of size
+            [num_rows, dim_emb_j].
+        index (Union[List[int], slice]): a list of row indices or a slice to
+            apply to each tensor in tensor_list.
```
Comment (weihua916) [inline test/data/test_multi_embedding_tensor.py:34]: A list of...

File: torch_frame/data/multi_embedding_tensor.py [hunk 5]
```diff
@@ -38,7 +38,10 @@ class MultiEmbeddingTensor(_MultiTensor):
     """
     def validate(self):
         assert self.offset[0] == 0
-        assert len(self.offset) == self.num_cols + 1
+        assert len(
+            self.offset) == self.num_cols + 1 or self.values.numel() == 0
```
[the remark this answers is not in the document]
Comment (weihua916) [reply torch_frame/data/multi_embedding_tensor.py:42]: Why can't we let `len(self.offset) == self.num_cols + 1` for regardless of whether the tensor is empty?

File: torch_frame/data/multi_embedding_tensor.py (hunk 2 above)
Comment (weihua916) [reply torch_frame/data/multi_embedding_tensor.py:218]: I don't think row indexing (`dim=0`) operation should change the offset, since `offset` here purely specifies the column slicing information.

File: torch_frame/data/multi_embedding_tensor.py [hunk 6]
```diff
@@ -98,29 +101,159 @@ def __getitem__(
...
+            end_idx: int = self._normalize_index(
+                index.stop or num_data,
+                dim=dim,
+                is_slice_end=True,
+            )
+            return self.narrow(
+                dim=dim,
+                start=start_idx,
+                length=end_idx - start_idx,
+            )
```
Comment (weihua916) [inline torch_frame/data/multi_embedding_tensor.py:255]: Do you think `_slice`, `narrow`, `select`, `index_select` can be shared with `MultiNestedTensor` (in other words, we put them in the parent `_MultiTensor` class)? Here we only need to implement `_row_index_select`, `_col_index_select` and so on.

**Review by weihua916 — COMMENTED** [review]
LGTM in general, thanks!

i think we can put some general methods to parent class `_MultiTensor` to allow more code sharing with `MultiNestedTensor`.

File: torch_frame/data/multi_embedding_tensor.py (hunk 6 above)
Comment (akihironitta) [reply torch_frame/data/multi_embedding_tensor.py:255]: Sounds good. I'll update the PR accordingly

File: torch_frame/data/multi_embedding_tensor.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (akihironitta) [reply torch_frame/data/multi_embedding_tensor.py:42]: Because we still keep the size (incl num_cols) for an empty instance. What would you suggest in this case?

[the remark this answers is not in the document]
Comment (weihua916) [reply torch_frame/data/multi_embedding_tensor.py:42]: For empty instance, we can still have positive num_cols and associated offset?

[the remark this answers is not in the document]
Comment (akihironitta) [reply torch_frame/data/multi_embedding_tensor.py:42]: Ah I might be misunderstanding something. I'll double check this once I'm back at computer. Thank you for your thorough review as always!

[the remark this answers is not in the document]
Comment (akihironitta) [reply torch_frame/data/multi_embedding_tensor.py]: Fixed!

File: torch_frame/data/multi_embedding_tensor.py (hunk 6 above)
Comment (akihironitta) [reply torch_frame/data/multi_embedding_tensor.py:256]: Actually, could I follow up on the refactor in a separate PR to first unblock #194 and #206? I'd like to keep this PR easy to review and not touch `MultiNestedTensor` if you're fine.

File: torch_frame/data/multi_embedding_tensor.py [hunk 7]
```diff
@@ -152,6 +285,60 @@ def _row_index_select(self, index: Tensor) -> 'MultiEmbeddingTensor':
...
+        torch.cumsum(new_col_dims, dim=0, out=offset[1:])
+        value_index = [
+            torch.arange(
+                self.offset[i],
+                self.offset[i + 1],
+                dtype=torch.long,
+                device=self.device,
+            ) for i in index
+        ]
+        value_index = torch.cat(value_index, dim=0)
```
Comment (weihua916) [inline torch_frame/data/multi_embedding_tensor.py:312]: Can you try using `batched_arange` to avoid for-loop? See my usage in `MultiNestedTensor`.

File: torch_frame/data/multi_embedding_tensor.py [hunk 8]
```diff
@@ -152,6 +285,60 @@ def _row_index_select(self, index: Tensor) -> 'MultiEmbeddingTensor':
...
+                num_cols=self.num_cols,
+                values=self.values[index].view(1, -1),
+                offset=self.offset,
+            )
+        elif dim == 1:
+            return MultiEmbeddingTensor(
+                num_rows=self.num_rows,
+                num_cols=1,
+                values=self.values[:,
+                                   self.offset[index]:self.offset[index + 1]],
```
Comment (weihua916) [inline torch_frame/data/multi_embedding_tensor.py:338]: super nit: a bit more beautiful indent?

File: torch_frame/data/multi_embedding_tensor.py (hunk 6 above)
Comment (weihua916) [reply torch_frame/data/multi_embedding_tensor.py:256]: Sounds good. Thanks.

**Review by weihua916 — APPROVED** [review]
LGTM after avoiding for-loop via `batched_arange` (see [here](https://github.com/pyg-team/pytorch-frame/blob/master/torch_frame/data/multi_nested_tensor.py#L492-L518)). Thank you.
